Images are only tagged after all blobs are uploaded

joaopapereira opened the pull request:
The images were being tagged after all the images (that were part of a bundle) were uploaded to the registry. Nevertheless, this was happening before the creation of the Images Location image. This could cause some concurrency problems if someone tried to pull the images as soon as it was tagged.

Fixes #211

joaopapereira committed:
Add Len function to ProcessedImages

joaopapereira committed:
Ensure images are only tagged after all blobs are uploaded

The images were being tagged after all the images were uploaded to the
registry. Nevertheless, this was happening before the creation of the
Images Location image. This could cause some concurrency problems if
someone tried to pull the images as soon as it was tagged.

cppforlife commented on pkg/imgpkg/cmd/copy_repo_src.go:
```diff
@@ -242,3 +252,60 @@ func imageRefDescriptorsMediaTypes(ids *imagedesc.ImageRefDescriptors) []string
+
+				switch {
+				case item.Image != nil:
+					err = c.registry.WriteTag(uploadOriginalTagRef, item.Image)
```
hmm. should we be doing this instead? https://github.com/vmware-tanzu/carvel-kbld/blob/develop/pkg/kbld/registry/registry.go#L118 (seems a bit weird to pass in entire image/index in there)

joaopapereira replied:
You need to provide ggcr with a `remote.Taggable` which is implemented by the `v1.Image` and `v1.ImageIndex` if you see that code from `kbld` there is a `remote.Get` call (https://github.com/vmware-tanzu/carvel-kbld/blob/develop/pkg/kbld/registry/registry.go#L130) that we do not need to do again since we already have a reference to a `remote.Taggable`.

If we did move to a similar interface as we do in `kbld` we would have to retrieve the Image that we already have from the registry. Do you think we need the extra step?

cppforlife replied:
oh doh. i totally glanced over that bit.

cppforlife commented on pkg/imgpkg/cmd/copy_repo_src.go:
```diff
@@ -242,3 +252,60 @@ func imageRefDescriptorsMediaTypes(ids *imagedesc.ImageRefDescriptors) []string
+					panic(fmt.Sprintf("Internal consistency: %s should be a digest", item.DigestRef))
+				}
+
+				uploadOriginalTagRef := digest.Tag(item.Tag)
```
uploadOriginalTagRef -> customTagRef

joaopapereira replied:
fixed

joaopapereira replied:
fixed

cppforlife commented on pkg/imgpkg/cmd/copy_repo_src.go:
```diff
@@ -242,3 +252,60 @@ func imageRefDescriptorsMediaTypes(ids *imagedesc.ImageRefDescriptors) []string
+	for _, item := range processedImages.All() {
+		item := item // copy
+
+		if item.Tag != "" {
```
might be better to do `item.Tag == "" continue` to avoid extra indention.

joaopapereira replied:
I make sense, fixed

joaopapereira replied:
this one is redundant

cppforlife commented on pkg/imgpkg/cmd/copy_repo_src.go:
```diff
@@ -242,3 +252,60 @@ func imageRefDescriptorsMediaTypes(ids *imagedesc.ImageRefDescriptors) []string
+	importThrottle := util.NewThrottle(c.Concurrency)
+
+	totalThreads := 0
+	errChVerifyImages := make(chan error, processedImages.Len())
```
errChVerifyImages -> errCh

joaopapereira replied:
fixed

joaopapereira replied:
this one is redundant.

cppforlife commented on pkg/imgpkg/cmd/copy_repo_src.go:
```diff
@@ -242,3 +252,60 @@ func imageRefDescriptorsMediaTypes(ids *imagedesc.ImageRefDescriptors) []string
+	return checkForAnyAsyncErrors(totalThreads, errChVerifyImages)
+}
+
+func checkForAnyAsyncErrors(numAsyncCalls int, errCh chan error) error {
```
i would favor inlining since its a pretty common pattern to happen after for-loop + goroutines

joaopapereira replied:
I was trying to keep the same pattern we already have in the code in a different place, but for now we only need it in one spot so it make sense to do what you suggest

joaopapereira replied:
This one is redundant

cppforlife commented on pkg/imgpkg/cmd/copy_repo_src.go:
```diff
@@ -242,3 +252,60 @@ func imageRefDescriptorsMediaTypes(ids *imagedesc.ImageRefDescriptors) []string
+				case item.Image != nil:
+					err = c.registry.WriteTag(uploadOriginalTagRef, item.Image)
+					if err != nil {
+						errChVerifyImages <- fmt.Errorf("Importing image as %s: %s", digest.Name(), err)
```
"Importing image as" -> "Tagging image"

joaopapereira replied:
fixed

cppforlife commented on pkg/imgpkg/cmd/copy_repo_src.go:
```diff
@@ -242,3 +252,60 @@ func imageRefDescriptorsMediaTypes(ids *imagedesc.ImageRefDescriptors) []string
 }
+
+func (c CopyRepoSrc) tagAllImages(processedImages *ctlimgset.ProcessedImages) error {
+	importThrottle := util.NewThrottle(c.Concurrency)
```
importThrottle -> throttle

joaopapereira replied:
fixed

cppforlife commented on pkg/imgpkg/cmd/copy_repo_src_test.go:
```diff
@@ -439,6 +439,31 @@ func TestToRepoBundleContainingANestedBundle(t *testing.T) {
+
+		for _, log := range *requestLog {
+			fmt.Printf("%s %s\n", log.Method, log.URL)
+		}
```
keep or not to keep?

joaopapereira replied:
Left over from debugging the test. Removed

cppforlife commented on pkg/imgpkg/cmd/copy_repo_src.go:
```diff
@@ -96,6 +101,11 @@ func (c CopyRepoSrc) CopyToRepo(repo string) (*ctlimgset.ProcessedImages, error)
+	}
+
 	informUserToUseTheNonDistributableFlagWithDescriptors(
 		c.ui, c.IncludeNonDistributable, imageRefDescriptorsMediaTypes(ids))
```
should we consolidate this into single exit point (instead of 2 currently)?

joaopapereira replied:
I was able to refactor the function and consolidate both the exit points, it should be good now

joaopapereira committed:
Refactor CopyToRepo

Merge both exit points of the function together
Remove the output parameter from Relocate that is no longer used
